src/blockchain_db/lmdb/db_lmdb.cpp

```diff
@@ -1907,6 +1969,61 @@ uint64_t BlockchainLMDB::get_block_timestamp(const uint64_t& height) const
   return ret;
 }
 
+std::vector<uint64_t> BlockchainLMDB::get_block_cumulative_rct_outputs(const std::vector<uint64_t> &heights) const
+{
+  LOG_PRINT_L3("BlockchainLMDB::" << __func__);
+  check_open();
+  std::vector<uint64_t> res;
+  int result;
+
+  if (heights.empty())
+    return {};
+  for (size_t i = 1; i < heights.size(); ++i)
+    if (heights[i] < heights[i-1])
+      throw0(DB_ERROR("Invalid heights array"));
+  res.reserve(heights.size());
+
+  TXN_PREFIX_RDONLY();
+  RCURSOR(rct_distribution);
+
+  MDB_stat db_stats;
+  if ((result = mdb_stat(m_txn, m_blocks, &db_stats)))
+    throw0(DB_ERROR(lmdb_error("Failed to query m_blocks: ", result).c_str()));
+  if (heights.back() >= db_stats.ms_entries)
+    throw0(BLOCK_DNE(std::string("Attempt to get rct distribution from height " + std::to_string(heights.back()) + " failed -- block size not in db").c_str()));
+
+  MDB_val k, data;
+  result = mdb_cursor_get(m_cur_rct_distribution, &k, &data, MDB_FIRST);
+  if (result)
+    throw0(DB_ERROR(lmdb_error("Error attempting to retrieve rct distribution from the db: ", result).c_str()));
+
+  uint64_t prev_height = m_rct_distribution_offset;
+  for (uint64_t height: heights)
+  {
+    if (height < m_rct_distribution_offset)
+    {
+      res.push_back(0);
+      continue;
+    }
+
+    size_t steps = height - prev_height;
+    while (steps--)
```

Comment:
Using the existing one wastes memory and wrecks caching when gathering the distribution.
I guess the pros of using height/value tuples are high though, though it does also waste memory. Not that much in the grand scheme of things.

Reply:
Though if we move to reusing the existing table, that makes height height free.
Using the existing table means we lose a bit less than 1 MB, which is not very much I suppose.

Reply:
So, in short:

- separate table: no wasted memory, good caching, slow to get to a single element
- reusing existing table: wasted memory, bad caching, ~instant lookup of a single element

Reply:
> The data is sorted: it's a cumulative count, so can never go down.

The caller is required to sort the values, but this is not documented in the header. Adding this information to the header is more brittle than taking the `std::vector` by value, and doing a sort internally to ensure correctness. Any caller can do a `std::move` to elide the copy entirely. Or a `if (!std::is_sorted(begin, end)) throw ....;` to enforce constraints (although arguable doing the sort is preferable).

I suppose one issue is that the `std::vector` is going to be pretty large, and thus either call is going to add lots of ops ... ? If so, the problem is likely with the API.

> Though if we move to reusing the existing table, that makes height height free.
Using the existing table means we lose ~6.4 MB.

I think its 9.3MiB of extra space to include it in the existing entry. I just took block height `1220517 * 8` for that value.

> separate table: 
> * no wasted memory, good caching, slow to get to a single element
> * reusing existing table: wasted memory, bad caching, ~instant lookup of a single element

What about the case where a `struct {height, value}` is stored as a new LMDB table? Is this function expected to _mostly_ retrieve the **entire** range of values? If so, then this implementation is reasonable. However, the `std::vector` being returned and provided as an argument will either continue to grow (somewhat quickly) as the number of blocks increases OR callers will grab a smaller range thus making the linear search inefficient. Since you commented elsewhere that an RPC for retrieving specific values is going to be added (and since the RPC is already allowed to restrict a range), using the extra space for a proper LMDB key makes sense.